// @deltazeroproduction/f1-udp-parser/build/src/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WEATHER = exports.TYRES = exports.TRACKS = exports.TEAMS = exports.SURFACES = exports.SESSION_TYPES = exports.SAFETY_CAR_STATUSES = exports.RULESETS = exports.PENALTIES = exports.PACKET_SIZES = exports.PACKETS = exports.NATIONALITIES = exports.INFRINGEMENTS = exports.GAME_MODES = exports.FUEL_MIX = exports.FORMULAS = exports.EVENT_CODES = exports.ERS_DEPLOY_MODE = exports.AI_DRIVER_NAMES = exports.DRIVERS = exports.BUTTON_FLAGS = void 0;
const buttonFlags_1 = require("./buttonFlags");
Object.defineProperty(exports, "BUTTON_FLAGS", { enumerable: true, get: function () { return buttonFlags_1.BUTTON_FLAGS; } });
const drivers_1 = require("./drivers");
Object.defineProperty(exports, "DRIVERS", { enumerable: true, get: function () { return drivers_1.DRIVERS; } });
Object.defineProperty(exports, "AI_DRIVER_NAMES", { enumerable: true, get: function () { return drivers_1.AI_DRIVER_NAMES; } });
const ersDeployMode_1 = require("./ersDeployMode");
Object.defineProperty(exports, "ERS_DEPLOY_MODE", { enumerable: true, get: function () { return ersDeployMode_1.ERS_DEPLOY_MODE; } });
const eventCodes_1 = require("./eventCodes");
Object.defineProperty(exports, "EVENT_CODES", { enumerable: true, get: function () { return eventCodes_1.EVENT_CODES; } });
const formulas_1 = require("./formulas");
Object.defineProperty(exports, "FORMULAS", { enumerable: true, get: function () { return formulas_1.FORMULAS; } });
const fuelMix_1 = require("./fuelMix");
Object.defineProperty(exports, "FUEL_MIX", { enumerable: true, get: function () { return fuelMix_1.FUEL_MIX; } });
const gameModes_1 = require("./gameModes");
Object.defineProperty(exports, "GAME_MODES", { enumerable: true, get: function () { return gameModes_1.GAME_MODES; } });
const infringements_1 = require("./infringements");
Object.defineProperty(exports, "INFRINGEMENTS", { enumerable: true, get: function () { return infringements_1.INFRINGEMENTS; } });
const nationalities_1 = require("./nationalities");
Object.defineProperty(exports, "NATIONALITIES", { enumerable: true, get: function () { return nationalities_1.NATIONALITIES; } });
const packets_1 = require("./packets");
Object.defineProperty(exports, "PACKETS", { enumerable: true, get: function () { return packets_1.PACKETS; } });
const packetSizes_1 = require("./packetSizes");
Object.defineProperty(exports, "PACKET_SIZES", { enumerable: true, get: function () { return packetSizes_1.PACKET_SIZES; } });
const penalties_1 = require("./penalties");
Object.defineProperty(exports, "PENALTIES", { enumerable: true, get: function () { return penalties_1.PENALTIES; } });
const rulesets_1 = require("./rulesets");
Object.defineProperty(exports, "RULESETS", { enumerable: true, get: function () { return rulesets_1.RULESETS; } });
const safetyCarStatuses_1 = require("./safetyCarStatuses");
Object.defineProperty(exports, "SAFETY_CAR_STATUSES", { enumerable: true, get: function () { return safetyCarStatuses_1.SAFETY_CAR_STATUSES; } });
const sessionTypes_1 = require("./sessionTypes");
Object.defineProperty(exports, "SESSION_TYPES", { enumerable: true, get: function () { return sessionTypes_1.SESSION_TYPES; } });
const surfaces_1 = require("./surfaces");
Object.defineProperty(exports, "SURFACES", { enumerable: true, get: function () { return surfaces_1.SURFACES; } });
const teams_1 = require("./teams");
Object.defineProperty(exports, "TEAMS", { enumerable: true, get: function () { return teams_1.TEAMS; } });
const tracks_1 = require("./tracks");
Object.defineProperty(exports, "TRACKS", { enumerable: true, get: function () { return tracks_1.TRACKS; } });
const tyres_1 = require("./tyres");
Object.defineProperty(exports, "TYRES", { enumerable: true, get: function () { return tyres_1.TYRES; } });
const weather_1 = require("./weather");
Object.defineProperty(exports, "WEATHER", { enumerable: true, get: function () { return weather_1.WEATHER; } });
